### DISCUSSION TITLE
Loading indicator recommendations and feedback

### DISCUSSION THREAD

**Discussion by basham** [indiana-university/rivet-source#543] [Feature and enhancement requests]
Over the last month, my team has been updating our use of loading indicators in the Student Appointment Scheduler. That lead me to question the use of [Rivet loading indicators](https://rivet.iu.edu/components/overlays/loading-indicator/). This [first CodePen](https://codepen.io/basham/pen/NWvKojN) outlines some problems with the current Rivet implementation. This [second CodePen](https://codepen.io/basham/pen/rNzOvvR) takes what was learned from the first and means to provide the start of a new recommendation. Both of these have been reviewed by the design community, but this should be also reviewed by the accessibility and Rivet teams. What feedback do you have on what was found and the new recommendations?

**Comment by basham** [comment]
This article is another good source for a particular problem with the loading indicator.

https://adrianroselli.com/2022/04/accessible-description-exposure.html#Background

> The ARIA 1.3 draft helpfully tells us [which roles cannot be assigned an accessible name through ARIA](https://w3c.github.io/aria/#namefromprohibited), which also applies to accessible descriptions. For example, the generic role is one of them. The generic role maps to <div> and <span>, meaning you cannot assign an aria-describedby to either element (nor an aria-label or aria-labelledby) unless it has an appropriate role.

The current recommendation for loading indicators is this implementation:

```html
<div class="rvt-loader" aria-label="Content loading"></div>
```

However, because there is no explicit or explicit role on this `div`, the `aria-label` will not work.

https://w3c.github.io/aria/#generic

**Reply by basham** [reply]
I recommend doing something like this:

```html
<div class="rvt-loader"></div>
<div class="rvt-sr-only">Content loading</div>
```

Or:

```html
<div class="rvt-loader">
  <div class="rvt-sr-only">Content loading</div>
</div>
```

**Comment by gjthomas** [comment]
Our team definitely got flagged from some a11y testing by using the example code in the Rivet 1.x docs. So, can confirm what is being said here and we will need to go back and fix the syntax in several places in our code. Would definitely recommend the example in the documents get changed.

**Reply by basham** [reply]
Was this testing your team did or another team? How did you discover the issue?

**Reply by gjthomas** [reply]
I'd say both.

For a little background, a lot of our tools launch with a loading screen, then the content appears when it is ready.

We had some testing done by another team that was exclusively using a screen reader and there was some confusion on what was happening. This also lead to try to get better automated testing from our QA side (via axe-core-cucumber https://github.com/dequelabs/axe-core-gems/tree/develop/packages/axe-core-cucumber ) and the rivet example in one of tools said:

```
1) aria-allowed-attr: Elements must only use allowed ARIA attributes (serious)
https://dequeuniversity.com/rules/axe/4.4/aria-allowed-attr?application=axeAPI

The following 1 node violate this rule:
Selector: #tool_content, #loading > .rvt-loader.rvt-loader--sm[aria-label="Content loading"]
HTML: <div class="rvt-loader rvt-loader--sm" aria-label="Content loading"></div>
Fix all of the following:
    - aria-label attribute cannot be used on a div with no valid role attribute.
```

This was a pretty recent discovery.
